### DISCUSSION TITLE
Sequencing work done by asynchronous processes

### DISCUSSION THREAD

**Discussion by ktbarrett** [cocotb/cocotb#2562] [Q&A] [answered]
Has anyone tried to solve the problem of providing synchronous interfaces to testbenching components in a general way?

If you don't know what I mean, let me explain. Most people seem to prefer writing their tests sequentially ("write to a system register, then read from some memory"), but most testbenching components need to be written in a concurrent way ("one process is pulling read requests off a queue, applying it to the interface, and sending it to an outstanding request queue to be matched up to the read response being read from a different interface by another process"). The best course is to do both. Use concurrency for testbenching components and blocking functions for the test. The real difficulty comes from bridging the two approaches. Right now, a test just sends Transactions to the Components via Queues. This is what UVM does as well (well they mostly attach "Sequences", AKA generators, to Components). However, Queues are asynchronous, it's kind of hard to sequence in a tests if you can't guarantee that any Transaction being sent out on a queue is "applied" strictly after another. I am looking for a way to provide that sequencing guarantee.

First, I tried making queues capable of synchronous operation by changing the semantics of a blocking put to block until the transaction was received by the consumer, rather than when it simply made it's way onto the queue. This has some problems when the transaction in question filters through many processes before being applied to the interface. It also doesn't handle when a transaction is broken into multiple smaller transaction.

Associating an Event with every Transaction which is set once the Transaction is applied to the Interface has issues with the last point as well, but can be solved by using a generalized semaphore-like event (fires once count number of `set()`s occurs). But it comes back to "is the transaction complete only when all pieces of a transaction are applied?"

Deciding when blocking finishes to enforcing sequencing depends entirely on what the current and the next action is. For example, applying AXI Full write transactions back-to-back doesn't need to wait, but if the next thing you intend on doing is triggering an interrupt that starts processing the data sent in that AXI transaction, then you need to block until you get a response. This makes the problem seem like it isn't generalizable. It would require either A. the processes to understand the user's intent and set an Event accordingly, which seems impossible, or B. require the user to understand the structure and role of each process in a system, require each process to report its state, and have the user decide when "done" is by combining that information; which totally erases any possibility of encapsulation and is a lot of work.

I have been trying to solve this problem for a while, but still haven't found a good solution. How do you decide that a Transaction has been applied and you can move on to the next thing? How do other verification methodologies like UVM and OSVVM handle this, if at all? My bet is that this is simply not a solvable problem and I'm beating my head against a 10m thick concrete wall thinking it's just drywall.

Sorry that this is a very vague and conceptual question.

**Comment by eric-wieser** [comment]
Can you come up with a simple concrete example of where you run into this problem? My impression is that the best way to handle this kind of thing is to convert any type of asynchronous transaction into a single async function, even if it happens on two different interfaces.

**Reply by ktbarrett** [reply]
For AXI Full Master we have as a suite of components. We have AXI Stream Drivers that have input and output Queues depending on channel directionality. We have AXI Full Master Driver that has a write request, read request, write response, and read request queues. There are two processes that takes each read and write requests and splits them out into individual transactions for each channel, and we have two processes that build up the response transactions from the individual channel transactions.

The user would like a `read(trans: AXIFullReadRequest) -> AXIFullReadResponse` and `write(trans: AXIFullWriteRequest) -> AXIFullWriteResponse` method. This is an "easy" examples since the blocking should occur until the response is seen, which comes out on a channel. 

A more difficult variant is the version of AXI Full used by Xilinx a lot where there is no write response channel. In that case `write` should block until the transaction has been applied on the interface. The blocking has no output queue to be based on, it's "internal". Additionally, there is both AW and multiple W channel transactions sent out per write request.

I'm not sure if there is a way to provide this blocking functionality "generally" to more than just these interfaces. It seems like this functionality is going to have to be special made per interface, which is a lot of work... I was hoping there was some feature to make this easy to provide or even allow the user to customize.

**Reply by benbr8** [reply]
I think what you want is already there, it's just a matter of architecture. You can get what you want quite easily by changing architecture of the AXI module from object oriented where one master handles all requests by pulling from a queue to a functional one, where for each transaction a task is forked or awaited. Since all requests share the same interface, you need to handle arbitration at a lower level, however this can be encapsulated and hidden from the user.

**Reply by ktbarrett** [reply]
> object-oriented

It's CSP, concurrency.

Sure for an AXI Full Master I could forgo a queue based approach, but eventually the buck stops. You will eventually need to cross from a typical (asynchronous) call stack into a concurrent system (that's exactly what HDLs are). Drivers sometimes need pre-assert, post-assert, and idle behavior and need to be an independent always running process. Monitors are the same way, they need to be an independent always running process. There is no way to bridge the gap between unrelated processes in a general way without queue-based communication. So you will eventually run into this problem.

You seem to be suggesting to make everything an asymmetric coroutine until you can't and then do some custom logic to implement the last layer. At least for Drivers.

**Reply by benbr8** [reply]
Well you can solve your issue with the AXI Master like that, at least ;). As long you have a physical interface which can only handle transactions in sequence you will at some point need to bring everything together.

If things need to be executed in sequence it would probably be easiest to express them as such in a sequential task which requires the infrastructure (drivers, monitors, ...) to expose the necessary interfaces as awaitable. Idle behaviour and the like do not necessarily conflict with that, depending on complexity.

The alternative is as you described, imposing snychronous behaviour on a design pattern which inherently is not. I can't imagine that this can be realized without a lot design overhead.

**Reply by ktbarrett** [reply]
Actually I just thought of a way. Drivers apply Transactions to Interfaces, Monitors observe Transactions on Interfaces and emit them. I have a generalized solution to Flow Control on Channels (they have blocking put/get and I expose top-level control of enable/disable of send/receive on Channels). I can disable communication in one Driver until I can use a Monitor to observe the Transaction making it onto the bus, then I can re-enable communication. That is a way to force sequencing without having to design a new mechanism for it and hardcode something different for every Driver.

Thank you beer, truly an inspiration to us all =)

BTW, I'm excited about `rstb` and potential collaborations between it and cocotb. Feel free to send me, any of the maintainers, or the cocotb mailing list reflections, opinions, suggestions, etc. I don't hear enough about cocotb from people who have used it seriously on a day-to-day basis.

**Reply by benbr8** [reply]
> BTW, I'm excited about `rstb` and potential collaborations between it and cocotb. Feel free to send me, any of the maintainers, or the cocotb mailing list reflections, opinions, suggestions, etc. I don't hear enough about cocotb from people who have used it seriously on a day-to-day basis.

Well that could also be because people are very content with how this project is progressing ;). Rstb actually had an optional Python front end at some point, but supporting that needs to be postponed, because there is still a lot of core functionality missing.

**Comment by ktbarrett** [comment, marked as answer]
As stated in https://github.com/cocotb/cocotb/discussions/2562#discussioncomment-870143, when it comes to sequencing application of Transaction on Interfaces, the easiest and general approach is to blindly send out Transactions while turning off the rest of the system, then re-enable the system once that Transaction is Monitored to have made it onto the bus/made it to it's destination/caused a response/whatever your criteria are.
